File: src/memory/builder/story_paths.py
"""Shared resolve-then-create path logic for Builder roadmap story packages.

Both Ariad Expand (``lifecycle.expand_delivery_story``) and the CLI's Plan /
checkpoint artifact-path lookups (``cli.build._canonical_package_path``) need
to answer the same question: given a roadmap item code, where does its
authored ``index.md`` package live on disk?

The single source of truth for an EXISTING package is its own heading
(``# <code> — <title>``), never a folder-naming convention or arithmetic on
the code+title -- an authored package can live under any folder name a human
chose; what makes it "CV2.DS1" is its heading, exactly as
``pull_candidates.py`` already resolves Pull candidates. ``resolve_story_directory``
is the one resolver every call site shares, so they can never diverge again.
This is the actual fix for the class of defect first seen as CR048 (DS6.TS5)
and recurring as CV22.DS7: Expand used to *derive* a path from code+title
arithmetic instead of resolving the real, already-authored package.

Creating a brand-new package (``create_story_directory``) is a different,
lower-stakes concern: since nothing is authored yet for that code, a new path
cannot diverge from or duplicate anything. Naming a not-yet-materialized
parent coordinate degrades gracefully -- resolve it if it already has a
package, else name it from the root roadmap snapshot
(``docs/project/roadmap/index.md``, which is the intentional planning record
for CVs before they have their own folder), else fall back to a bare code
segment. This never fails loud, and it never uses "/"-joined title-chain
arithmetic across code levels -- the previous design's other defect, since
the delivery cursor only ever carries a story's own leaf title, never a full
ancestor-title chain.
"""

from __future__ import annotations

from pathlib import Path

from memory.builder.pull_candidates import inspect_roadmap_snapshot
from memory.builder.roadmap_grammar import HEADING_RE, is_legacy_path
from memory.utils import kebab_slug


class StoryPackageAmbiguityError(ValueError):
    """More than one roadmap package's heading claims the same code."""


def title_leaf(title: str) -> str:
    """Return the final ``/``-separated segment of a title, trimmed.

    Roadmap titles are sometimes given as a full ancestor chain (``"CV title
    / DS title"``); this extracts just the item's own leaf segment. A plain,
    non-chain title passes through unchanged.
    """
    return title.split("/")[-1].strip()


def _sanitize_code_segment(code: str) -> str:
    """Filesystem-safe rendering of a work-item code for folder derivation.

    Dots are the code's own level separator (``CV2.DS1``), so they convert to
    hyphens first, preserving the existing folder-naming convention for
    well-formed codes unchanged. Everything else routes through
    ``kebab_slug`` -- the same sanitizer already applied to titles -- so an
    unsanitized ``/`` or ``..`` surviving in a candidate-table code cell
    (Navigator- or LLM-authored, unlike a folder-name title) can no longer
    create unintended nested directories within the roadmap tree (closes
    D-012).
    """
    return kebab_slug(code.replace(".", "-"))


def story_folder_name(code: str, title: str) -> str:
    """Canonical folder name for a story: ``<code>-<title-slug>``."""
    code_part = _sanitize_code_segment(code)
    slug = kebab_slug(title)
    return f"{code_part}-{slug}" if slug else code_part


def resolve_story_directory(project_path: Path, code: str) -> Path | None:
    """Return the directory of the authored package whose ``index.md``
    heading code equals ``code`` exactly, or ``None`` if no package claims it.

    Raises ``StoryPackageAmbiguityError`` if more than one package does -- the
    roadmap-integrity CI guard (duplicate-heading check, see
    ``find_duplicate_roadmap_headings`` and ``docs_lint.py``) exists precisely
    to catch this before it ever reaches runtime resolution.
    """
    roadmap_root = (project_path / "docs" / "project" / "roadmap").resolve()
    if not roadmap_root.is_dir():
        return None
    matches = _group_roadmap_headings(roadmap_root).get(code, [])
    if not matches:
        return None
    if len(matches) > 1:
        raise StoryPackageAmbiguityError(
            f"{len(matches)} roadmap packages claim code {code!r}: "
            + ", ".join(str(path) for path in matches)
        )
    return matches[0]


def _group_roadmap_headings(roadmap_root: Path) -> dict[str, list[Path]]:
    """Group every non-legacy ``docs/**/index.md`` file under a code by its
    own heading (``# <code> — <title>``). Shared scan used by both
    ``resolve_story_directory`` (single-code lookup) and
    ``find_duplicate_roadmap_headings`` (whole-tree duplicate detection) so
    the two can never drift apart.
    """
    by_code: dict[str, list[Path]] = {}
    for index_path in sorted(roadmap_root.rglob("index.md")):
        if is_legacy_path(index_path, roadmap_root):
            continue
        try:
            content = index_path.read_text(encoding="utf-8")
        except OSError:
            continue
        heading = HEADING_RE.search(content)
        if heading:
            by_code.setdefault(heading.group("code").strip(), []).append(index_path.parent)
    return by_code


def find_duplicate_roadmap_headings(project_path: Path) -> dict[str, tuple[Path, ...]]:
    """Return every code claimed by more than one ``docs/project/roadmap/**/index.md``
    heading, mapped to its claiming directories.

    This is the CI-facing, whole-tree counterpart to ``resolve_story_directory``'s
    per-code ambiguity guard -- the static check (see ``docs_lint.py``) that keeps
    the "one code -> one package" invariant from ever silently breaking again
    (the class of defect first seen as CR048 / DS6.TS5, recurring as CV22.DS7).
    Returns an empty dict when the roadmap root does not exist or every code is
    unique.
    """
    roadmap_root = (project_path / "docs" / "project" / "roadmap").resolve()
    if not roadmap_root.is_dir():
        return {}
    grouped = _group_roadmap_headings(roadmap_root)
    return {code: tuple(paths) for code, paths in grouped.items() if len(paths) > 1}


def _parent_code(code: str) -> str | None:
    """``'CV2.DS1'`` -> ``'CV2'``; ``'DS-35'`` -> ``None`` (no dotted parent)."""
    if "." not in code:
        return None
    return code.rsplit(".", 1)[0]


def _bare_code_segment(code: str) -> str:
    return _sanitize_code_segment(code)


def _snapshot_title(project_path: Path, code: str) -> str | None:
    """Best-effort title for ``code`` from the root roadmap index's own
    compact snapshot (handles both the legacy CV-heading grammar and the
    CV/DS table grammar -- see ``pull_candidates._snapshot_items_from_content``).
    Used only to NAME a not-yet-materialized parent coordinate; never used to
    decide where an EXISTING package lives (that is always
    ``resolve_story_directory``).
    """
    snapshot = inspect_roadmap_snapshot(project_path, journey="", method="ariad")
    for item in snapshot.items:
        if item.code == code:
            return item.title
    return None


def _parent_directory(project_path: Path, code: str) -> Path:
    """Best available directory to nest a NEW ``code`` under: resolve an
    existing authored package by heading; else name it from the root roadmap
    snapshot; else recurse to the grandparent and fall back to a bare code
    segment. Never fails loud -- nothing is authored yet for ``code``, so
    nothing can diverge; this only affects the readability of a brand-new
    folder name, never correctness or safety.
    """
    resolved = resolve_story_directory(project_path, code)
    if resolved is not None:
        return resolved
    roadmap_root = (project_path / "docs" / "project" / "roadmap").resolve()
    parent = _parent_code(code)
    base = roadmap_root if parent is None else _parent_directory(project_path, parent)
    title = _snapshot_title(project_path, code)
    folder = story_folder_name(code, title) if title else _bare_code_segment(code)
    return base / folder


def create_story_directory(project_path: Path, code: str, leaf_title: str) -> Path:
    """Canonical directory for a NEW package, nested under its best-available
    parent coordinate. Only the story's own leaf title slugs its own folder
    name -- title chains are never used for arithmetic across code levels.
    """
    roadmap_root = (project_path / "docs" / "project" / "roadmap").resolve()
    parent = _parent_code(code)
    base = roadmap_root if parent is None else _parent_directory(project_path, parent)
    target = (base / story_folder_name(code, title_leaf(leaf_title))).resolve()
    if not target.is_relative_to(roadmap_root):
        raise ValueError(f"story directory escapes roadmap root: {target}")
    return target

File: ts/src/builder/storyPaths.ts
// CV22.DS7.US8 plateau 1 — resolve-then-create for roadmap story packages.
//
// Port of `src/memory/builder/story_paths.py`.
//
// The rule that matters: an EXISTING package is located by its own heading
// (`# <code> — <title>`), never by a folder-naming convention or by arithmetic
// on code+title. An authored package can live under any folder a human chose,
// and what makes it `CV2.DS1` is its heading. Python learned this twice —
// CR048 in DS6.TS5, recurring in CV22.DS7 — so the resolver is shared by every
// call site rather than re-derived per caller.
//
// Creating a NEW package is the lower-stakes half: nothing is authored for that
// code yet, so a new path cannot duplicate or contradict anything, and naming a
// not-yet-materialized parent degrades gracefully instead of failing.
//
// Confinement is not incidental. `createStoryDirectory` builds a path from a
// caller-supplied code and title — candidate-table cells, which may be
// Navigator- or model-authored — so the final path is checked to be inside the
// roadmap root and the code segment is sanitized before it can create nested
// directories (Python's D-012).

import { relative, resolve, sep } from "node:path";
import { kebabSlug } from "#util/slug.ts";
import { inspectRoadmapSnapshot, roadmapPaths } from "./pullCandidates.ts";
import { matchHeading } from "./roadmapGrammar.ts";
import { readRoadmapFile, scanRoadmapIndexFiles } from "./roadmapScan.ts";

/** Python `StoryPackageAmbiguityError`: two packages claim one code. */
export class StoryPackageAmbiguityError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "StoryPackageAmbiguityError";
  }
}

/**
 * Python `title_leaf`: the final `/`-separated segment, stripped.
 *
 * Roadmap titles are sometimes authored as an ancestor chain
 * (`"CV title / DS title"`), so this takes the item's own segment. The cost is
 * that a title legitimately containing a slash — `"Builder/Ariad tree"` — is
 * read as a chain and loses its head. That is Python's behavior and this story's
 * own Pull surface shows it; reproduced here, recorded as a CR.
 */
export function titleLeaf(title: string): string {
  const segments = title.split("/");
  return (segments[segments.length - 1] ?? "").trim();
}

/**
 * Python `_sanitize_code_segment`: dots are the code's own level separator, so
 * they become hyphens FIRST (preserving `cv2-ds1` folder naming for well-formed
 * codes), and everything else goes through the same slug sanitizer titles use.
 * That is what stops a `/` or `..` in a code cell from creating directories.
 */
function sanitizeCodeSegment(code: string): string {
  return kebabSlug(code.replaceAll(".", "-"));
}

/** Python `story_folder_name`: `<code>-<title-slug>`, or the code alone. */
export function storyFolderName(code: string, title: string): string {
  const codePart = sanitizeCodeSegment(code);
  const slug = kebabSlug(title);
  return slug ? `${codePart}-${slug}` : codePart;
}

/** Every code claimed by a non-legacy `index.md` heading, mapped to its directories. */
function groupRoadmapHeadings(roadmapRoot: string): Map<string, string[]> {
  const byCode = new Map<string, string[]>();
  for (const file of scanRoadmapIndexFiles(roadmapRoot)) {
    const content = readRoadmapFile(file.absolutePath);
    if (content === null) continue;
    const heading = matchHeading(content);
    if (!heading) continue;
    const code = heading.code.trim();
    const directory = file.components.slice(0, -1).join("/");
    const existing = byCode.get(code);
    if (existing) existing.push(directory);
    else byCode.set(code, [directory]);
  }
  return byCode;
}

/**
 * Python `resolve_story_directory`: the directory of the authored package whose
 * heading code equals `code`, `null` when none does, and a throw when more than
 * one does.
 *
 * Returns a path relative to the roadmap root's parent chain as components
 * joined with `/`; callers that need an absolute path join it to the root.
 */
export function resolveStoryDirectory(projectRoot: string, code: string): string | null {
  // ABSOLUTE, like Python's `(project_path / … ).resolve()`. This was relative until
  // CV22.DS7.US8 plateau 3, and the divergence was invisible because every test and
  // every real caller passed an absolute project path: the front door reads it from
  // the journey row, where `journey set-path` stores an absolute one. A
  // project-relative path made Python print absolute package paths and TypeScript
  // print relative ones in the same surface, and — worse — would have compared a
  // relative candidate against a resolved root inside `createStoryDirectory`'s
  // confinement guard.
  const roadmapRoot = resolve(roadmapPaths(projectRoot).roadmapRoot);
  const matches = groupRoadmapHeadings(roadmapRoot).get(code) ?? [];
  if (matches.length === 0) return null;
  if (matches.length > 1) {
    const paths = matches.map((directory) =>
      directory ? `${roadmapRoot}${sep}${directory.split("/").join(sep)}` : roadmapRoot,
    );
    throw new StoryPackageAmbiguityError(
      `${matches.length} roadmap packages claim code '${code}': ${paths.join(", ")}`,
    );
  }
  const only = matches[0] ?? "";
  return only ? `${roadmapRoot}${sep}${only.split("/").join(sep)}` : roadmapRoot;
}

/** Python `find_duplicate_roadmap_headings`: the CI-facing whole-tree counterpart. */
export function findDuplicateRoadmapHeadings(projectRoot: string): Map<string, string[]> {
  const { roadmapRoot } = roadmapPaths(projectRoot);
  const duplicates = new Map<string, string[]>();
  for (const [code, directories] of groupRoadmapHeadings(roadmapRoot)) {
    if (directories.length > 1) duplicates.set(code, directories);
  }
  return duplicates;
}

/** Python `_parent_code`: `CV2.DS1` -> `CV2`; `DS-35` -> `null`. */
function parentCode(code: string): string | null {
  if (!code.includes(".")) return null;
  return code.slice(0, code.lastIndexOf("."));
}

/** Python `_snapshot_title`: name a not-yet-materialized coordinate from the root snapshot. */
function snapshotTitle(projectRoot: string, code: string): string | null {
  const snapshot = inspectRoadmapSnapshot(projectRoot, { journey: "", method: "ariad" });
  for (const item of snapshot.items) {
    if (item.code === code) return item.title;
  }
  return null;
}

/**
 * Python `_parent_directory`: resolve an existing package by heading, else name
 * it from the root snapshot, else recurse to the grandparent and fall back to a
 * bare code segment. Never fails loud — nothing is authored for the code yet.
 */
function parentDirectory(projectRoot: string, code: string): string {
  const resolved = resolveStoryDirectory(projectRoot, code);
  if (resolved !== null) return resolved;
  // Resolved for the same reason as above: Python's `_parent_directory` bases every
  // derived path on the resolved roadmap root.
  const roadmapRoot = resolve(roadmapPaths(projectRoot).roadmapRoot);
  const parent = parentCode(code);
  const base = parent === null ? roadmapRoot : parentDirectory(projectRoot, parent);
  const title = snapshotTitle(projectRoot, code);
  const folder = title ? storyFolderName(code, title) : sanitizeCodeSegment(code);
  return `${base}${sep}${folder}`;
}

/**
 * Python `create_story_directory`: the canonical directory for a NEW package,
 * nested under its best-available parent coordinate, with only the story's own
 * leaf title slugging its own folder.
 *
 * Throws when the resolved target escapes the roadmap root. That guard is the
 * reason a `../` code or an absolute-looking title cannot write outside
 * `docs/project/roadmap/`, and it is checked AFTER resolution so a symlink or a
 * `..` surviving sanitization is still caught.
 */
export function createStoryDirectory(projectRoot: string, code: string, leafTitle: string): string {
  const { roadmapRoot } = roadmapPaths(projectRoot);
  const resolvedRoadmapRoot = resolve(roadmapRoot);
  const parent = parentCode(code);
  const base = parent === null ? resolvedRoadmapRoot : parentDirectory(projectRoot, parent);
  const target = resolve(base, storyFolderName(code, titleLeaf(leafTitle)));
  const relation = relative(resolvedRoadmapRoot, target);
  const escapes =
    relation.startsWith("..") || relation.startsWith(`${sep}`) || resolve(relation) === relation;
  if (escapes) {
    throw new Error(`story directory escapes roadmap root: ${target}`);
  }
  return target;
}
